x/stakeibc/keeper/icacallbacks_delegate_test.go

```diff
@@ -0,0 +1,270 @@
+	records := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx())
+	s.Require().Len(records, 0, "number of deposit records")
+}
+
+func (s *KeeperTestSuite) TestDelegateCallback_DelegateCallbackTimeout() {
+	tc := s.SetupDelegateCallback()
+	validArgs := tc.validArgs
+	// a nil ack means the request timed out
+	validArgs.ack = nil
+	err := stakeibckeeper.DelegateCallback(s.App.StakeibcKeeper, s.Ctx(), validArgs.packet, validArgs.ack, validArgs.args)
+	s.Require().NoError(err)
+
+	// Confirm stakedBal has not increased
+	hostZone, found := s.App.StakeibcKeeper.GetHostZone(s.Ctx(), chainId)
+	s.Require().True(found)
+	s.Require().Equal(int64(tc.initialState.stakedBal), int64(hostZone.StakedBal), "stakedBal should not have increased")
+
+	// Confirm deposit record has been removed
+	records := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx())
+	s.Require().Len(records, 1, "number of deposit records")
+	record := records[0]
+	s.Require().Equal(recordtypes.DepositRecord_STAKE, record.Status, "deposit record status should be timeout")
+}
+
+func (s *KeeperTestSuite) TestDelegateCallback_DelegateCallbackErrorOnHost() {
+	tc := s.SetupDelegateCallback()
+	validArgs := tc.validArgs
+	// an error ack means the tx failed on the host
+	fullAck := channeltypes.Acknowledgement{Response: &channeltypes.Acknowledgement_Error{Error: "error"}}
+	validArgs.ack = &fullAck
+
+	err := stakeibckeeper.DelegateCallback(s.App.StakeibcKeeper, s.Ctx(), validArgs.packet, validArgs.ack, validArgs.args)
+	s.Require().NoError(err)
+
+	// Confirm stakedBal has not increased
+	hostZone, found := s.App.StakeibcKeeper.GetHostZone(s.Ctx(), chainId)
+	s.Require().True(found)
+	s.Require().Equal(int64(tc.initialState.stakedBal), int64(hostZone.StakedBal), "stakedBal should not have increased")
+
+	// Confirm deposit record has been removed
+	records := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx())
+	s.Require().Len(records, 1, "number of deposit records")
+	record := records[0]
+	s.Require().Equal(recordtypes.DepositRecord_STAKE, record.Status, "deposit record status should be timeout")
+}
+
+func (s *KeeperTestSuite) TestDelegateCallback_WrongCallbackArgs() {
+	tc := s.SetupDelegateCallback()
+	validArgs := tc.validArgs
+	
+	err := stakeibckeeper.DelegateCallback(s.App.StakeibcKeeper, s.Ctx(), validArgs.packet, validArgs.ack, []byte("random bytes"))
+	s.Require().EqualError(err, "unexpected EOF")
+
+	// Confirm stakedBal has not increased
+	hostZone, found := s.App.StakeibcKeeper.GetHostZone(s.Ctx(), chainId)
+	s.Require().True(found)
+	s.Require().Equal(int64(tc.initialState.stakedBal), int64(hostZone.StakedBal), "stakedBal should not have increased")
+
+	// Confirm deposit record has been removed
+	records := s.App.RecordsKeeper.GetAllDepositRecord(s.Ctx())
+	s.Require().Len(records, 1, "number of deposit records")
+	record := records[0]
+	s.Require().Equal(recordtypes.DepositRecord_STAKE, record.Status, "deposit record status should be timeout")
```

Comment:
for future reference, you could probably bundle all these into a function like
`CheckStateIfCallbackFailed`. Can be a little tedious to have to modify every one of these failure cases if something changes

Reply:
much cleaner, done.